### PR TITLE
Bugfix: ignore index + macro combination in classification metrics

### LINKED ISSUE #2441: Error in computing `MulticlassRecall` macro average when `ignore_index` is specified
## 🐛 Bug

In `MulticlassRecall` the `ignore_index` argument leads to incorrect behaviour when computing the macro average. It seems as if the metric still divides by the total number of classes rather than ignoring the `ignore_index` class (while treating the `ignore_index` class as having a score of `0`).

### To Reproduce

```python
import torch
from torchmetrics.classification.precision_recall import MulticlassRecall

metric = MulticlassRecall(num_classes=2, ignore_index=0, average="macro")
y_true = torch.tensor([0, 0, 1, 1])
y_pred = torch.tensor([
    [0.9, 0.1],  # Correct
    [0.9, 0.1],  # Correct
    [0.9, 0.1],  # Incorrect
    [0.1, 0.9],  # Correct
])
metric.update(y_pred, y_true)
print(metric.compute())  # Prints 0.25, but I would expect the result to be 0.5
```

### Expected behavior

I would expect the recall score to be `0.5`.

In the toy example above I have two classes (0 and 1) and two instances of each class. I want to ignore class 0 altogether (hence `ignore_index=0`). The first prediction for class 1 (the third prediction overall) is incorrect (it predicts class 0 with confidence 0.9 instead), but the other prediction is correct. Therefore, I would expect the recall score to be `0.5`. 

However, the actual result is `0.25`. It looks as if this metric is still dividing by the total number of classes when computing the macro average instead of ignoring the `ignore_index` class (while treating the `ignore_index` class as having a score of `0`).

(Notice that the predictions for class 0 are in fact correct. If I leave out the `ignore_index=0` altogether, I get an expected result of `0.75`.)

### Environment

- TorchMetrics version: 1.3.1 (installed via pip)
- Python version: 3.10
- PyTorch version: 2.2.1
- OS: Ubuntu 22.04

### PR DESCRIPTION
**Pull request by SkafteNicki**
## What does this PR do?

Fixes #2441

<details>
  <summary>Before submitting</summary>

- [ ] Was this **discussed/agreed** via a Github issue? (no need for typos and docs improvements)
- [ ] Did you read the [contributor guideline](https://github.com/Lightning-AI/torchmetrics/blob/master/.github/CONTRIBUTING.md), Pull Request section?
- [ ] Did you make sure to **update the docs**?
- [ ] Did you write any new **necessary tests**?

</details>

<details>
  <summary>PR review</summary>

Anyone in the community is free to review the PR once the tests have passed.
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

</details>

## Did you have fun?

Make sure you had fun coding 🙃

### CODE REVIEW

**Comment by Borda** [comment]
@SkafteNicki, what is missing here to make it land? :chipmunk:

**Comment by DimitrisMantas** [comment]
Is this still happening?

**Comment by JeroenMandersloot** [comment]
Any chance that this will be worked on in the near future?

**Comment by Borda** [comment]
> Any chance that this will be worked on in the near future?

cc: @SkafteNicki

**Comment by Borda** [comment]
> Any chance that this will be worked on in the near future?

do you want to take it over and reopen @SkafteNicki commits as part of your/new PR so you will share contribution? 🦩

**Comment by Borda** [comment]
@DimitrisMantas @JeroenMandersloot are you interested in taking this over and finishing it? 🦩

**Comment by Borda** [comment]
@rittik9, would you be interested in finishing this one? 🐰

**Comment by Borda** [comment]
@SkafteNicki let's finish it or close it? 🐿️

**Comment by robmarkcole** [comment]
Bumping @SkafteNicki
